### DISCUSSION TITLE
Show custom form or filter row above header widgets and link them with header widgets

### DISCUSSION THREAD

**Discussion by FrazeColder** [filamentphp/filament#10100] [Help] [answered]
I am using Filament PHP v3 and have created a custom page. The content of that page presents two header widgets. One line chart and one table widget. 

Above those two widgets, I want to display a row with different filter options such as start date, end date, etc. 

However, I am having difficulties to archive my goal. I have already tried to use `HasFiltersForm` as well `HasFormSchema` but I am not able to display any forms or filters to input data. 

This is my custom page
```
class ShowClicks extends Page
{
    protected static string $resource = ShortUrlClickResource::class;

    protected static string $view = 'filament.resources.short-url-click-resource.pages.show-clicks';

    protected function getHeaderWidgets(): array
    {
        return [
            ShortUrlClicksChart::make(),
            MostShortUrlClicks::make(),
        ];
    }
}
```

When I was using `HasFiltersForm` I have added the following function to my code ([source][1]):
```
ublic function filtersForm(Form $form): Form
    {
        return $form
            ->schema([
                Section::make()
                    ->schema([
                        DatePicker::make('startDate'),
                        DatePicker::make('endDate'),
                        // ...
                    ])
                    ->columns(3),
            ]);
    }
```

But there was no filter or form displayed... The same goes for `InteractsWithForms` ([source][2]): 
```
protected function getFormSchema()
    {
        return [
            Forms\Components\TextInput::make('site_name'),
        ];
    }
```

I don't see any form/inputable filter to use on my custom page...

How can I add a row above my header widgets to display filters which interact with my widgets?

Kind regards

  [1]: https://filamentphp.com/docs/3.x/panels/dashboard#filtering-widget-data
  [2]: https://github.com/filamentphp/filament/discussions/1169#discussioncomment-1930945

**Comment by leandrocfe** [comment, marked as answer]
What about a header action with a custom form?
https://filamentphp.com/docs/3.x/actions/modals#modal-forms

**Reply by FrazeColder** [reply]
And how can I make the filter inputs available in the widgets?

**Reply by leandrocfe** [reply]
Via events?

https://livewire.laravel.com/docs/events#dispatching-events

You could also try to pass properties 
https://filamentphp.com/docs/3.x/panels/resources/widgets/#passing-properties-to-widgets-on-resource-pages

**Reply by FrazeColder** [reply]
Oh, thank you! I would like to go with the passing properties option.

Will the data be automatically be passed to the widgets when they change or do I have to listen for them within the widgets?

**Reply by leandrocfe** [reply]
Honestly, I haven't tried it but I think this will be auto updated

**Reply by FrazeColder** [reply]
Alright, thank you. But I am not able to display a row of inputs/filters above my widgets.

I am just able to show an action button... How can I perform this with a custom form?

**Reply by leandrocfe** [reply]
You could create a custom widget
https://filamentphp.com/docs/3.x/panels/resources/widgets/#creating-a-resource-widget

This widget would be the first with a columnSpan = 'full'

You could create a custom livewire component with a form

https://filamentphp.com/docs/3.x/panels/resources/widgets/#creating-a-resource-widget

Then, render the livewire component in the widget view

```html
<div>
@livewire('custom-form')
</div>
```
